text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+# Handling of Equivocation Catastrophes
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0000](https://github.com/casperlabs/ceps/pull/0000)
+
+This CEP describes a proposal for the process of handling equivocation catastrophes in Highway consensus (or, more generally: handling situations that break the assumptions behind the security of the consensus algorithm being used).
+
+## Motivation
+
+[motivation]: #motivation
+
+Every consensus algorithm's safety guarantees rely on some assumptions being upheld. In the case of Highway, the main such assumption is about the validators not equivocating, that is: not producing conflicting votes. The algorithm can withstand some amount of equivocations, but if the total weight of equivocating validators exceeds a threshold called the Fault Tolerance Threshold (FTT), they could revert finalized blocks, which could lead for example to double spending. We call such a situation an "equivocation catastrophe" and it is something that cannot be handled automatically (as it would effectively contradict well-established results about the limitations of consensus algorithms in general). Thus, we require some kind of an external mechanism for dealing with such situations, should they ever happen.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The procedure for handling the equivocation catastrophes will have to consist of two main parts: what the software is expected to do in such a case, and what are the manual actions required to proceed.
+
+The obvious first thing that should happen in the case of an equivocation catastrophe is that the software should alert the users about the situation. The node software is aware of the FTTs of the blocks, so when it detects equivocations, it can notice if the total weight of the equivocators exceeds the FTT of a block. When the FTT of any block is exceeded, the users should be made aware of it.
```

Comment:
@afck, does this imply option 1 needs serious clarification regarding the meaning of FTT? E.g., does it mean a _still_ finalized block, or the last finalized block before the catastrophe was observed (and by whom?)?

Reply:
It definitely has to mean a _still_ finalized block, because as you say, the "last observed" one is subjective.

But, if 40% equivocated, should we take the last block with 41% FTT? Or would we want some safety margin and require, say, 45%? Or, if we have a 33% FTT configured, would we now want 33% _of the remaining 60%_, which would mean an absolute FTT of 40% + 20% = 60%?
That's something we could also make some recommendation for.

Reply:
Presumably we are ejecting the equivocating validators in a genuine (non-bug-related) equivocation attack, so it should be something that would be finalized if we excise them? "Highest ranked with highest mass among remaining validators" perhaps?

Reply:
Yes, that should be equivalent with the remaining 60% above. So:
* If we go with the last block with 41% FTT, we basically assume that none of the remaining 60% has equivocated. (If we're all in a Zoom call and agree exactly on the set of equivocators, that's probably reasonable.)
* If we go with the last block with 60% FTT, that means requiring a 33% FTT _among the remaining 60%_, i.e. it would mean 33% if the equivocators had never existed.